flagscale/train/models/deepseek_v3/multi_token_predictor.py

```diff
@@ -0,0 +1,293 @@
+    """
+
+    def __init__(
+        self,
+        config: TransformerConfig,
+        transformer_layer_spec: ModuleSpec,
+        vocab_size: int,
+        max_sequence_length: int,
+        position_embedding_type: Literal['learned_absolute', 'rope', 'none'] = 'learned_absolute',
+        scatter_embedding_sequence_parallel: bool = True,
+        parallel_output: bool = True,
+        share_embeddings_and_output_weights: bool = True,
+        embedding_activation_buffer: Optional[List[torch.Tensor]] = None,
+        grad_output_buffer: Optional[List[torch.Tensor]] = None,
+    ):
+        super().__init__(config=config)
+
+        self.config = config
+        self.num_mtp_predictor = config.num_mtp_predictor
+        
+        self.mtp_modules = torch.nn.ModuleList([
+            DeepSeekMultiTokenPredictorLayer(
+                    config=self.config,
+                    transformer_layer_spec=transformer_layer_spec,
+                    vocab_size=vocab_size,
+                    max_sequence_length=max_sequence_length,
+                    position_embedding_type=position_embedding_type,
+                    scatter_embedding_sequence_parallel=scatter_embedding_sequence_parallel,
+                    parallel_output=parallel_output,
+                    share_embeddings_and_output_weights=share_embeddings_and_output_weights,
+                    embedding_activation_buffer=embedding_activation_buffer,
+                    grad_output_buffer=grad_output_buffer,
+            ) for i in range(self.num_mtp_predictor)
+        ])
+
+    def forward(
+        self,
+        input_ids: Tensor,
+        position_ids: Tensor,
+        attention_mask: Tensor,
+        pre_hidden_states: Tensor,
+        decoder_input: Tensor = None,
+    ) -> Tensor:
+        """Forward pass of the multi token prediction module.
+
+        Args:
+            input_ids (Tensor): The input tokens or input embeddings
+            pre_hidden_states (Tensor): The hidden states from previous multi token prediction module or main model
+
+        Returns:
+            Tensor: The output logits
+        """
+        
+        # init input mask
+        if decoder_input is not None:
+            s, b, _ = decoder_input.shape
+        else:
+            b, s = input_ids.shape
+        input_mask = torch.zeros(s, b).unsqueeze(2).cuda().type(torch.bool)
+        input_mask = roll_input_mask(input_mask)
+        
+        logits_mtps = []
+        for i in range(self.num_mtp_predictor):
+            logits_mtp, pre_hidden_states = self.mtp_modules[i](
+                input_ids=input_ids,
+                position_ids=position_ids,
+                attention_mask=attention_mask,
+                pre_hidden_states=pre_hidden_states,
+                decoder_input=decoder_input,
+                input_mask=input_mask,
+            )
+            logits_mtps.append(logits_mtp.transpose(0, 1).contiguous())
+            input_mask = roll_input_mask(input_mask)
```

Comment:
As shown below:
![image](https://github.com/user-attachments/assets/21a157dd-a6c3-4075-b15a-50baa2597322)
The input sequences should be `[t1,t2,t3,t4], [t2,t3,t4,t5]` not `[t1,t2,t3,t4, t5], [t2,t3,t4,t5]` and `attention_mask` should keep consistent with `input_ids`